src/rng.cpp

```diff
@@ -7,40 +7,70 @@
 #include <cstdlib>
 #include <random>
 
-long rng( long val1, long val2 )
+static std::mt19937 &rng_get_engine();
+static std::uniform_int_distribution<long> rng_int_dist;
+static std::uniform_real_distribution<double> rng_real_dist;
+static std::normal_distribution<double> rng_normal_dist;
+
+uintmax_t rng_bits()
 {
-    long minVal = ( val1 < val2 ) ? val1 : val2;
-    long maxVal = ( val1 < val2 ) ? val2 : val1;
-    return minVal + long( ( maxVal - minVal + 1 ) * double( rand() / double( RAND_MAX + 1.0 ) ) );
+    size_t size = sizeof( uintmax_t ) * CHAR_BIT;
+    size_t word = rng_get_engine().word_size;
+    size_t shift = 0;
+    div_t res = div( size, word );
+    if( res.rem ) {
+        res.quot++;
+    }
+    res.quot--;
+    uintmax_t ret = 0;
+    while( 1 ) {
+        ret |= uintmax_t( rng_get_engine()() ) << shift;
+        if( res.quot > 0 ) {
+            shift += word;
+            res.quot--;
+            continue;
+        }
+        break;
+    }
+    return ret;
 }
 
-double rng_float( double val1, double val2 )
+long rng( long lo, long hi )
 {
-    double minVal = ( val1 < val2 ) ? val1 : val2;
-    double maxVal = ( val1 < val2 ) ? val2 : val1;
-    return minVal + ( maxVal - minVal ) * double( rand() ) / double( RAND_MAX + 1.0 );
+    if( lo > hi ) {
+        std::swap( lo, hi );
+    }
+
+    decltype( rng_int_dist.param() ) range( lo, hi );
+    rng_int_dist.param( range );
+    return rng_int_dist( rng_get_engine() );
 }
 
-bool one_in( int chance )
+double rng_float( double lo, double hi )
 {
-    return ( chance <= 1 || rng( 0, chance - 1 ) == 0 );
+    if( lo > hi ) {
+        std::swap( lo, hi );
+    }
+
+    decltype( rng_real_dist.param() ) range( lo, hi );
+    rng_real_dist.param( range );
+    return rng_real_dist( rng_get_engine() );
 }
 
-//this works just like one_in, but it accepts doubles as input to calculate chances like "1 in 350,52"
-bool one_in_improved( double chance )
+bool one_in( long chance )
 {
-    return ( chance <= 1 || rng_float( 0, chance ) < 1 );
+    return ( chance <= 1 || rng( 0, chance - 1 ) == 0 );
 }
 
 bool x_in_y( double x, double y )
 {
-    return ( static_cast<double>( rand() ) / RAND_MAX ) <= ( static_cast<double>( x ) / y );
+    return rng_float( 0, 1 ) * y <= x;
 }
 
-int dice( int number, int sides )
+long dice( long number, long sides )
```

Comment:
Consistency's sake. What the function actually needed was unsigned arguments (and you probably want to be using `uint_leastN_t` types because that makes the most sense), because there's no reason for any of the parameters or return value to be negative, but hardly any part of the code base cares about writing correct code. Signed types are used willy-nilly like we're in Java land where everything is signed and wraps.

`long` is 32-bit on MSVC and 64-bit on Linux.

Reply:
Consistency with what? The rest of the project uses int, unless you have a demonstrable problem that's caused by using int here, this is more likely to break things than to fix anything.

Reply:
Consistency with the rest of the random number functions (they use long).

Besides, this makes no difference on MSVC (`sizeof (long) == sizeof (int)`). The only time this is going to cause issues is when the code using the `dice` function was broken to begin with - a dice roll so big that it would have overflowed an `int` (which may still be a problem with the current code). As such there is no reason to deny changing it here.

None of the (mostly trivial) uses of this function do anything crazy with large values - changing the signature to `long`s will not cause any bugs in reasonable code.